### PR TITLE
feat(parachain): add types

### PR DESCRIPTION
**Pull request by kanishkatn**
## Changes

Added few parachain types needed for #3241 

Breaking this out so that the subsystems could use this.

## Tests

```sh
go test lib/parachain
```

## Issues

#3186 

## Primary Reviewer

@kishansagathiya

### CODE REVIEW

File: lib/parachain/types_test.go
Comment (jimjbrettj) [inline lib/parachain/types_test.go:332]: I think we talked about this before but wanted to bring up again, is it worth having long hex strings like this in a separate file for readability?

**Review by jimjbrettj — APPROVED** [review]
small question, but overall looks good!

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go:332]: Yes, thank you! I feel it doesn't affect the readability since it is just one line. 

I believe there was also an open conversation regarding the block number type.

File: lib/parachain/types.go
Comment (dimartiro) [inline lib/parachain/types.go:21]: I think we can move this to `dot/types/block.go`

**Review by dimartiro — APPROVED** [review]
LGTM, I left only one suggestion to have all block types in the same file

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:21]: Thanks @dimartiro !

I believe we're looking to contain the types to its respective package. 

@timwu20 What do you think?

**Review by edwardmack — APPROVED** [review]
This looks great.

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:68]: this looks some comment polkadots folks added for their reference. we don't need this.

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:86]:
```suggestion
```

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:245]: I can't find corresponding rust structs, can you point me to them?

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:344]:
```suggestion
	CoreIndex        CoreIndex           `scale:"3"`
```

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:132]: add scale tags

File: lib/parachain/types.go
Comment (kishansagathiya) [reply lib/parachain/types.go:25]: honestly, having a separate type for block number is n't really necessary.
At most places we have just used uint32 directly.

If we do however want to add a separate type for this then `dot/types` is not a bad place, BlockNumber is something that could get used across the project.

File: lib/parachain/types_test.go
Comment (kishansagathiya) [inline lib/parachain/types_test.go:178]: this could be in one line

File: lib/parachain/types_test.go
Comment (kishansagathiya) [inline lib/parachain/types_test.go:198]: Do we need so many validatorIDs to see if encoding works well?

File: lib/parachain/types_test.go
Comment (kishansagathiya) [inline lib/parachain/types_test.go:236]: do we need this many AssignmentIDs?

File: lib/parachain/types_test.go
Comment (kishansagathiya) [inline lib/parachain/types_test.go:217]: do we need this many AuthorityDiscoveryIDs?

File: lib/parachain/types_test.go
Comment (kishansagathiya) [inline lib/parachain/types_test.go:44]: do we need this many validatorids?

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go:198]: Yes

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go:236]: Yes

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go:178]: You're right!

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:132]: Good catch!

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:344]: Good catch!

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:245]: These aren't used. I'm removing them

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:68]: Yes, I've been using it for my reference. I'll be cleaning these up before merging!

File: lib/parachain/types.go
Comment (timwu20) [reply lib/parachain/types.go:25]: If we don't have any receiver methods for a custom type, I would agree with Kishan, and probably just use the primitive.

Comment (kanishkatn) [reply lib/parachain/types.go:25]: What is a receiver method?

File: lib/parachain/types.go
Comment (timwu20) [inline lib/parachain/types.go:35]:
```suggestion
type GroupRotationInfo struct {
	// SessionStartBlock is the block number at which the session started
	SessionStartBlock BlockNumber `scale:"1"`
	// GroupRotationFrequency indicates how often groups rotate. 0 means never.
	GroupRotationFrequency BlockNumber `scale:"2"`
	// Now indicates the current block number.
	Now BlockNumber `scale:"3"`
}
```
You don't need to define the `scale` struct tags if the fields are already in the right order.  I put the scale tags support in just in case we decided to re-order the attributes based for whatever reason.

File: lib/parachain/types_test.go
Comment (timwu20) [reply lib/parachain/types_test.go:314]: I think you should use `go:embed` for things like this.  You can store them as test assets.

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:25]: I believe otherwise. It provides clearer intent and enhances code readability.

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:35]: @timwu20  Were you meant to remove the scale tags in the suggested change?

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go:314]: I am not opposed to it, but if we should use `go:embed`, I feel we should do that in a separate PR since there are quite a bit of already existing tests like this.

File: lib/parachain/types.go
Comment (timwu20) [reply lib/parachain/types.go:35]: I'm just letting you know it's not required, so I didn't make the suggestion.

Comment (kanishkatn) [reply lib/parachain/types.go:35]: haha, my bad. I thought I saw something else!

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go:314]: Issue created #3309

File: lib/parachain/types.go
Comment (axaysagathiya) [inline lib/parachain/types.go:211]: the type must be a pointer to `ValidationCode` as it's  `Option<ValidationCode>` type in Rust.
```suggestion
	// New validation code.
	NewValidationCode *ValidationCode `scale:"3"`
```

**Review by axaysagathiya — CHANGES_REQUESTED** [review]

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go:25]: Resolving this cause I need these types in dispute-coordinator. Since this is being merged to a feature branch, happy to continue the conversation in the future.

Comment (kanishkatn) [reply lib/parachain/types.go:25]: Well, it looks like I can't merge this. Is there anything I can do to get this PR merged so that I can continue my work on dispute-coordinator?

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:149]: remove

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:155]: remove

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:103]:
```suggestion
	// core. nil if there is nothing queued for this core or there is no possibility of timing
```

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:96]:
```suggestion
	// core, if any. nil if there is nothing queued for this core.
```

File: lib/parachain/types.go
Comment (kishansagathiya) [inline lib/parachain/types.go:16]: remove?

File: lib/parachain/types.go
Comment (timwu20) [reply lib/parachain/types.go:25]: > I believe otherwise. It provides clearer intent and enhances code readability. Could you tell me why you think having a receiver makes a difference?

That's the criteria I use when creating custom types.  Unless I want to add a method to the type, I'm ok with just using primitives.  I can go either way in this case, but if we were to use the parachain package and use types that use `parachain.BlockNumber`, the user would have to cast any existing primitive or custom types to `parachain.BlockNumber`, which could be avoided by just using the primitive.

File: lib/parachain/types_test.go
Comment (timwu20) [reply lib/parachain/types_test.go:314]: > I am not opposed to it, but if we should use `go:embed`, I feel we should do that in a separate PR since there are quite a bit of already existing tests like this.

Could we not just start with this test? This blob is quite large and makes it difficult to scroll through on the "files changed" tab.  This `types_test.go` file gets hidden as well I think due to the large blob.

Comment (timwu20) [reply lib/parachain/types_test.go:314]: > Issue created #3309

Thanks for creating this issue.  There are other cases in the codebase I'd like to move blobs into embedded assets.

File: lib/parachain/types.go
Comment (kanishkatn) [reply lib/parachain/types.go]: Keeping this in for reference https://github.com/ChainSafe/gossamer/pull/3297#discussion_r1214740051

File: lib/parachain/types_test.go
Comment (kanishkatn) [reply lib/parachain/types_test.go]: It's done!

**Review by axaysagathiya — APPROVED** [review]

**Pull request merged**
